### PR TITLE
mgr/dashboard: NFS Export form fixes

### PR DESCRIPTION
**Pull request by dnyanee1997**
Fixes: https://tracker.ceph.com/issues/67400
           https://tracker.ceph.com/issues/68105

Signed-off-by: Dnyaneshwari talwekar <dtalweka@redhat.com>

Tracker: https://bugzilla.redhat.com/show_bug.cgi?id=2303196
               https://bugzilla.redhat.com/show_bug.cgi?id=2269104
               https://bugzilla.redhat.com/show_bug.cgi?id=2302756

Steps: 
1.Navigate to Dashboard -> File -> NFS
2.Create NFS
3.On Create NFS page we can see "CephFS Path" as non-editable and Pseudo as blank field.

Updated Requirement:-
1. On Create NFS SubVolume selection should be required if SubVolume group is "_nogroup".
2. CephFS path should never be single "/"

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

https://github.com/user-attachments/assets/d58b8ad5-cc1e-4df5-918b-6363a62b919f

### CODE REVIEW

**Comment by cloudbehl** [comment]
@dnyanee1997 Congratulation on your first PR.

can you share a recording of the change and how its working

**Comment by dnyanee1997** [comment]
1. Navigate to Dashboard -> File -> NFS 
2. Create NFS
3. On Create NFS page we can see "**CephFS Path"** as non-eaditable and **Pseudo** as blank field.

**Comment by nizamial09** [comment]
thank you @dnyanee1997. this fixes the pseudo part but part of the issue is that its allowing to create an export on top of an volume which should not be the case. simply put the path should never be `/` but it should be a path of whatever inside `/` for eg: `/test_group/test_directory`.

A simple fix would be to make `_nogroup` as the default value for the subvolume group which will essentially mean that the `CephFS Path` will always be `/volumes/_nogroup` which could be what we want. let me know if you have any doubt.

https://github.com/user-attachments/assets/1cf7dd0e-142f-4ed6-95a5-b4ad3bbcd53e

**Comment by afreen23** [comment]
> thank you @dnyanee1997. this fixes the pseudo part but part of the issue is that its allowing to create an export on top of an volume which should not be the case. simply put the path should never be `/` but it should be a path of whatever inside `/` for eg: `/test_group/test_directory`.
> 
Yep agree 
> A simple fix would be to make `_nogroup` as the default value for the subvolume group which will essentially mean that the `CephFS Path` will always be `/volumes/_nogroup` which could be what we want. let me know if you have any doubt.

Sounds good but if user deletes the default value and tries to do "/" it would pass then, so we should have some validation.
We can have a custom validation that checks for isolated "/" but allows anything else like "/t/g"

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.html [hunk 1]
```diff
@@ -244,7 +244,8 @@
                    formControlName="path"
                    [ngbTypeahead]="pathDataSource"
                    (selectItem)="pathChangeHandler()"
-                   (blur)="pathChangeHandler()">
+                   (blur)="pathChangeHandler()"
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.html:247]: I am not understanding why we have added this here while removing it below ?

Moreover functions should not be removed but you may just add a validation

**Review by afreen23 — CHANGES_REQUESTED** [review]

**Comment by afreen23** [comment]
hey @dnyanee1997 ,

The bug is different and what are we doing here is completely opposite :/ 
Is that the BZ attached wrong ?

Thanks

**Comment by dnyanee1997** [comment]
Hi @afreen23, As per discussion with Ankush it was said that CephFS should be **editable** and bydefault subvolume group should be _nogroup.
cc@cloudbehl

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts [hunk 2]
```diff
@@ -544,23 +574,6 @@ export class NfsFormComponent extends CdForm implements OnInit {
...
-
-      if (_.isString(path) && !_.isEmpty(path)) {
-        newPseudo += '/' + path;
-      } else if (!_.isEmpty(this.nfsForm.getValue('fsal').user_id)) {
-        newPseudo += '/' + this.nfsForm.getValue('fsal').user_id;
-      }
-    }
-
-    return newPseudo;
-  }
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts:563]: Why delete this ?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts [hunk 3]
```diff
@@ -521,13 +558,6 @@ export class NfsFormComponent extends CdForm implements OnInit {
     );
   }
 
-  pathChangeHandler() {
-    if (!this.isEdit) {
-      this.nfsForm.patchValue({
-        pseudo: this.generatePseudo()
-      });
-    }
-  }
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts:530]: What is the reason to delete this ?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.html [hunk 4]
```diff
@@ -188,10 +188,7 @@
               <option *ngIf="allsubvolgrps === null"
                       value=""
                       i18n>Loading...</option>
-              <option *ngIf="allsubvolgrps !== null && allsubvolgrps.length === 0"
-                      value=""
-                      i18n>-- No CephFS subvolume group available --</option>
-              <option *ngIf="allsubvolgrps !== null && allsubvolgrps.length > 0"
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.html:194]: Why deleting this ?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.html [hunk 5]
```diff
@@ -205,7 +202,7 @@
            *ngIf="storageBackend === 'CEPH' && nfsForm.getValue('fsal').fs_name">
         <label class="cd-col-form-label"
                for="subvolume"
-               i18n>Subvolume</label>
+               i18n><span [ngClass]="{'required': nfsForm.getValue('subvolume_group') == defaultSubVolGroup}">Subvolume</span></label>
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.html:205]:
```suggestion
               [ngClass]="{'required': nfsForm.getValue('subvolume_group') == defaultSubVolGroup}"
               i18n>Subvolume</label>
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts [hunk 6]
```diff
@@ -258,8 +271,8 @@ export class NfsFormComponent extends CdForm implements OnInit {
       cluster_id: new UntypedFormControl('', {
         validators: [Validators.required]
       }),
-      path: new UntypedFormControl('/', {
-        validators: [Validators.required]
+      path: new UntypedFormControl('', {
+        validators: [Validators.required, this.isolatedSlashValidator()]
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts:275]: We have a "CdValidators.custom()", consider using that.

**Comment by afreen23** [comment]
> Hi @afreen23, As per discussion with Ankush it was said that CephFS should be **editable** and bydefault subvolume group should be _nogroup. cc@cloudbehl

I think the original BZ is https://bugzilla.redhat.com/show_bug.cgi?id=2302756 ?

**Review by afreen23 — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts [hunk 7]
```diff
@@ -181,11 +182,19 @@ export class NfsFormComponent extends CdForm implements OnInit {
...
   volumeChangeHandler() {
-    this.pathChangeHandler();
-    const fs_name = this.nfsForm.getValue('fsal').fs_name;
+   const fs_name = this.nfsForm.getValue('fsal').fs_name;
+    const subvolgrp = this.nfsForm.getValue('subvolume_group');
     this.getSubVolGrp(fs_name);
-  }
+    this.getSubVol();
 
+    if(subvolgrp == this.defaultSubVolGroup){
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts:190]:
```suggestion
    if(subvolgrp === this.defaultSubVolGroup){
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.html [hunk 8]
```diff
@@ -224,7 +221,10 @@
             <option *ngFor="let subvolume of allsubvols"
                     [value]="subvolume.name">{{ subvolume.name }}</option>
           </select>
-        </div>
+          <span *ngIf="nfsForm.getValue('subvolume_group') == 
+          defaultSubVolGroup && !nfsForm.getValue('subvolume')" class="invalid-feedback">SubVolume selection is required.
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.html:225]:
```suggestion
          <span *ngIf="nfsForm.getValue('subvolume_group') ===  defaultSubVolGroup && !nfsForm.getValue('subvolume')" 
                       class="invalid-feedback"
                       i18n>This field is required.
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.html [hunk 9]
```diff
@@ -251,7 +251,9 @@
             <span class="invalid-feedback"
                   *ngIf="nfsForm.showError('path', formDir, 'required')"
                   i18n>This field is required.</span>
-
+                  <span class="invalid-feedback"
+                  *ngIf="nfsForm.get('path').hasError('isolatedSlash') && nfsForm.get('path').touched"
+                  i18n>This field cannot be just a single "/".</span>
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.html:256]:
```suggestion
                  i18n>Export on CephFS volume not allowed.</span>
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts [hunk 10]
```diff
@@ -181,11 +182,19 @@ export class NfsFormComponent extends CdForm implements OnInit {
...
+      this.updatePath('/volumes/' + this.defaultSubVolGroup);
+      this.setUpVolumeValidation();
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts:192]: This whole logic can go in a separate function.
And reuse here and in `setSubVolGrpPath`

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts [hunk 11]
```diff
@@ -213,7 +223,7 @@ export class NfsFormComponent extends CdForm implements OnInit {
       if (subvolGroup === this.defaultSubVolGroup) {
         this.updatePath('/volumes/' + this.defaultSubVolGroup);
         resolve();
-      } else {
+      } else if(subvolGroup != ''){
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts:226]: We must add comments here as the logic in this file is complex.
Lets add some useful comments why we are doing this. Thanks

**Review by afreen23 — CHANGES_REQUESTED** [review]
Thanks @dnyanee1997 , some comments from my side.
I think by mistake you listed thsi BZ too here https://bugzilla.redhat.com/show_bug.cgi?id=2269104 which is not being fixed currently. Thanks

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts [hunk 12]
```diff
@@ -207,15 +207,12 @@ export class NfsFormComponent extends CdForm implements OnInit {
...
     return new Promise<void>((resolve, reject) => {
-      const subvolGroup = this.nfsForm.getValue('subvolume_group');
-      const fs_name = this.nfsForm.getValue('fsal').fs_name;
 
-      if (subvolGroup === this.defaultSubVolGroup) {
-        this.updatePath('/volumes/' + this.defaultSubVolGroup);
-        resolve();
-      } else {
+     const { fs_name, subvolgrp } = this.checkVolumeValidation();
+     resolve();
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts:212]: Shouldn't we be resolving it at the end?

**Review by afreen23 — CHANGES_REQUESTED** [review]
Hey @dnyanee1997 ,

Lets add some proper comments to code as well.

And looks like you might have lint errors and to fix that run `npm run fix` 
from `/frontend` dir.
Also some html ones might need manual fix so check them by `npm run lint`

Thanks

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.html [hunk 13]
```diff
@@ -193,6 +194,9 @@
         <option *ngFor="let subvolume of allsubvols"
                 [value]="subvolume.name">{{ subvolume.name }}</option>
       </cds-select>
+      <ng-template #subvolumeError><span *ngIf="nfsForm.getValue('subvolume_group') === 
+        defaultSubVolGroup && !nfsForm.getValue('subvolume')" class="invalid-feedback" i18n>This field is required.
```
[the remark this answers is not in the document]
Comment (dnyanee1997) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.html:198]: In the recent Carbon changes seems like we are following this for showing messages, took reference from it.Please correct me if its no required.

[the remark this answers is not in the document]
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.html:198]: yeah, we'll need the template and this template is passed to the `[invalidText]` attr of the input element. that's how the carbon does it.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.html [hunk 14]
```diff
@@ -220,6 +224,9 @@
           <span class="invalid-feedback"
                 *ngIf="nfsForm.showError('path', formDir, 'required')"
                 i18n>This field is required.</span>
+          <span class="invalid-feedback" 
+                *ngIf="nfsForm.get('path').hasError('isolatedSlash') && nfsForm.get('path').touched" 
+                i18n>Export on CephFS volume not allowed.</span>
```
[the remark this answers is not in the document]
Comment (dnyanee1997) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.html:229]: Replacing pre with "<code>/</code>" , as <pre>/</pre> having issue with spacing.
 with pre tag  -> Screenshot from 2024-09-04 15-50-26
with code ->
/home/dnyaneshwari/Pictures/Screenshots/Screenshot from 2024-09-04 15-51-54.png

File: src/pybind/mgr/dashboard/frontend/src/app/shared/constants/cephfs.constant.ts [hunk 15]
```diff
@@ -0,0 +1 @@
+export const DEFAULT_SUBVOLUME_GROUP = '_nogroup';
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/constants/cephfs.constant.ts:1]: nit: missing new line
```suggestion
export const DEFAULT_SUBVOLUME_GROUP = '_nogroup';

```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts [hunk 16]
```diff
@@ -207,27 +208,43 @@ export class NfsFormComponent extends CdForm implements OnInit {
...
+    this.getSubVolGrp(fsName);
+
+    if (subvolGroup === this.defaultSubVolGroup) {
+      this.updatePath('/volumes/' + this.defaultSubVolGroup);
+      this.setUpVolumeValidation();
+    } else {
+      this.updatePath('');
+    }
+
+    return { fsName, subvolGroup };
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts:245]: we can easily  get this value hence no need to return.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts [hunk 17]
```diff
@@ -181,9 +182,8 @@ export class NfsFormComponent extends CdForm implements OnInit {
   }
 
   volumeChangeHandler() {
-    this.pathChangeHandler();
-    const fs_name = this.nfsForm.getValue('fsal').fs_name;
-    this.getSubVolGrp(fs_name);
+    this.getSubVol();
+    this.checkVolumeValidation();
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts:186]: There is redundancy involved..
volumeChangeHandler() -> getSubVol() -> setSubVolGrpPath -> checkVolumeValidation

And then again checkVolumeValidation.
Either we remove it from here or change our logic flow.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts [hunk 18]
```diff
@@ -207,27 +208,43 @@ export class NfsFormComponent extends CdForm implements OnInit {
...
+        .subscribe(
+          (path) => {
+            this.updatePath(path);
+            resolve();
+          },
+          (error) => reject(error)
+        );
+      } else {
+        this.updatePath('');
+        this.setUpVolumeValidation();
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts:225]: iN checkVolumeValidation we are setting path and validation and here again, is it not redoing/overwriting ?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts [hunk 19]
```diff
@@ -207,27 +208,43 @@ export class NfsFormComponent extends CdForm implements OnInit {
...
       }
+      resolve();
     });
   }
 
+  // Checking if subVolGroup is "_nogroup" and updating path to default as "/volumes/_nogroup else blank."
+  checkVolumeValidation(): { fsName: string; subvolGroup: string } {
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts:232]:
```suggestion
  isDefaultSubvolumeGroup(): boolean {
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts [hunk 20]
```diff
@@ -207,27 +208,43 @@ export class NfsFormComponent extends CdForm implements OnInit {
...
+    const subvolGroup = this.nfsForm.getValue('subvolume_group');
+    const fsName = this.nfsForm.getValue('fsal').fs_name;
+
+    this.getSubVolGrp(fsName);
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts:236]: why this here ?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts [hunk 21]
```diff
@@ -207,27 +208,43 @@ export class NfsFormComponent extends CdForm implements OnInit {
...
+
+    if (subvolGroup === this.defaultSubVolGroup) {
+      this.updatePath('/volumes/' + this.defaultSubVolGroup);
+      this.setUpVolumeValidation();
+    } else {
+      this.updatePath('');
+    }
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts:243]:
```suggestion
      return subvolGroup === this.defaultSubVolGroup
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts [hunk 22]
```diff
@@ -207,27 +208,43 @@ export class NfsFormComponent extends CdForm implements OnInit {
...
   setSubVolGrpPath(): Promise<void> {
     return new Promise<void>((resolve, reject) => {
-      const subvolGroup = this.nfsForm.getValue('subvolume_group');
-      const fs_name = this.nfsForm.getValue('fsal').fs_name;
-
-      if (subvolGroup === this.defaultSubVolGroup) {
-        this.updatePath('/volumes/' + this.defaultSubVolGroup);
-        resolve();
-      } else {
+      const { fsName, subvolGroup } = this.checkVolumeValidation();
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts:211]: Here also we are updating path and below as well.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts [hunk 23]
```diff
@@ -340,6 +359,18 @@ export class NfsFormComponent extends CdForm implements OnInit {
       )
     });
   }
+ // SubVolume is required if SubVolume Group is "_nogroup".
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts:362]:
```suggestion
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts [hunk 24]
```diff
@@ -340,6 +359,18 @@ export class NfsFormComponent extends CdForm implements OnInit {
...
+  setUpVolumeValidation() {
+    const subvolumeGroup = this.nfsForm.get('subvolume_group').value;
+    const subVolumeControl = this.nfsForm.get('subvolume');
+
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts:366]:
```suggestion
// Subvolume is required if subvolume group is "_nogroup".
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts [hunk 25]
```diff
@@ -258,8 +274,11 @@ export class NfsFormComponent extends CdForm implements OnInit {
       cluster_id: new UntypedFormControl('', {
         validators: [Validators.required]
       }),
-      path: new UntypedFormControl('/', {
-        validators: [Validators.required]
+      path: new UntypedFormControl('', {
+        validators: [
+          Validators.required,
+          CdValidators.custom('isolatedSlash', this.isolatedSlashCondition) // Path can never be single "/".
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts:280]:
```suggestion
          CdValidators.custom('isIsolatedSlash', this.isolatedSlashCondition) // Path can never be single "/".
```

**Review by afreen23 — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts (hunk 19 above)
Comment (dnyanee1997) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/nfs/nfs-form/nfs-form.component.ts:232]: This function will be required as it has return value as fsName and subGroupValue.

**Comment by cloudbehl** [comment]
@dnyanee1997 tests are failing: https://jenkins.ceph.com/job/ceph-pull-requests/143611/console

**Review by cloudbehl — APPROVED** [review]
tested it, it works perfectly.

**Comment by dnyanee1997** [comment]
jenkins test make check

**Comment by dnyanee1997** [comment]
jenkins test dashboard cephadm

**Comment by dnyanee1997** [comment]
jenkins test dashboard

**Comment by dnyanee1997** [comment]
jenkins test make check

**Comment by dnyanee1997** [comment]
jenkins test make check arm64

**Pull request merged**